go/v1beta1/storage/elasticsearch.go

```diff
@@ -482,7 +482,174 @@ func (es *ElasticsearchStorage) CreateNote(ctx context.Context, projectId, noteI
 
 // BatchCreateNotes batch creates the specified notes in memstore.
 func (es *ElasticsearchStorage) BatchCreateNotes(ctx context.Context, pID, uID string, notes map[string]*pb.Note) ([]*pb.Note, []error) {
+	log := es.logger.Named("BatchCreateNotes")
+	log.Debug("creating notes")
+
+	var (
+		createdNotes   []*pb.Note
+		errs           []error
+		msearchReqBody bytes.Buffer
+		bulkBody       bytes.Buffer
+	)
+
+	newNotesMap := make(map[string]*pb.Note)
+
+	indexMetadataMsearch := &esMsearchQueryFragment{
+		Index: notesIndex(pID),
+	}
+
+	msearchMetadata, _ := json.Marshal(indexMetadataMsearch)
+	msearchMetadata = append(msearchMetadata, "\n"...)
+
+	for name := range notes {
+
+		noteName := fmt.Sprintf("projects/%s/notes/%s", pID, name)
+		newNotesMap[noteName] = notes[name]
+		searchBody := &esSearch{
+			Query: &filtering.Query{
+				Term: &filtering.Term{
+					"name": noteName,
+				},
+			},
+		}
+
+		data, err := json.Marshal(searchBody)
+		if err != nil {
+			return nil, []error{
+				createError(log, "error marshaling note search msearchReqBody", err),
+			}
+		}
+
+		dataBytes := append(data, "\n"...)
+		msearchReqBody.Grow(len(msearchMetadata) + len(dataBytes))
+		msearchReqBody.Write(msearchMetadata)
+		msearchReqBody.Write(dataBytes)
+	}
+
+	log.Debug("attempting ES multisearch", zap.String("payload", string(msearchReqBody.Bytes())))
+
+	res, err := es.client.Msearch(
+		bytes.NewReader(msearchReqBody.Bytes()),
+		es.client.Msearch.WithContext(ctx),
+	)
+
+	if err != nil {
+		return nil, []error{
+			createError(log, "failed while sending request to ES", err),
+		}
+	}
+	if res.IsError() {
+		return nil, []error{
+			createError(log, "unexpected response from ES", nil),
+		}
+	}
+
+	response := &esMsearch{}
+	err = decodeResponse(res.Body, response)
+	if err != nil {
+		return nil, []error{
+			createError(log, "error decoding ES response", nil),
+		}
+	}
+
+	msearchResponses := []esMsearchResponse{}
+	msearchResponses = response.Responses
+
+	for i := range msearchResponses {
+		if msearchResponses[i].Hits.Total.Value != 0 {
+			name := msearchResponses[i].Hits.Hits[0].Source.Name
+			errs = append(errs, createError(log, fmt.Sprintf("Note with the name %s already exists", name), nil))
+			delete(newNotesMap, name)
+		}
+	}
+
+	keys := make([]string, 0, len(newNotesMap))
+	for key := range newNotesMap {
+		keys = append(keys, key)
+	}
+
+	indexMetadataBulk := &esBulkQueryFragment{
+		Index: &esBulkQueryIndexFragment{
+			Index: notesIndex(pID),
+		},
+	}
+
+	bulkMetadata, _ := json.Marshal(indexMetadataBulk)
+	bulkMetadata = append(bulkMetadata, "\n"...)
+
+	// build the request bulkBody using newline delimited JSON (ndjson)
+	// each note is represented by two JSON structures:
+	// the first is the bulkMetadata that represents the ES operation, in this case "index"
```

Comment:
I think these comments read better when they aren't directly referencing variables, rather explaining how the request works in plain english.